File: src/test/testBroker.cpp
#include "../broker.h"

#include "./utility.h"

#include "gtest/gtest.h"
#include "pxr/pxr.h"
#include "pxr/usd/sdf/path.h"
#include "pxr/usd/usd/primRange.h"
#include "pxr/usd/usd/stage.h"

TEST(AddPrim, NoticesComparison)
{
    auto stage = PXR_NS::UsdStage::CreateInMemory();
    auto broker = PXR_NS::NoticeBroker::Create(stage);

    namespace _Broker = PXR_NS::UsdBrokerNotice;
    using _USD = PXR_NS::UsdNotice;

    // Subscribe to notices emitted by the broker.
    ::Test::Listener<
        _Broker::StageNotice,
        _Broker::StageContentsChanged,
        _Broker::ObjectsChanged,
        _Broker::StageEditTargetChanged,
        _Broker::LayerMutingChanged
    > listener1(stage);

    // Subscribe to Usd Notices emitted by the stage.
    ::Test::Listener<
        _USD::StageNotice,
        _USD::StageContentsChanged,
        _USD::ObjectsChanged,
        _USD::StageEditTargetChanged,
        _USD::LayerMutingChanged
    > listener2(stage);

    // Edit the stage...
    stage->DefinePrim(PXR_NS::SdfPath {"/Foo"});

    // Ensure that similar notices are received via the stage and the broker. 
    ASSERT_EQ(listener1.Received<_Broker::StageNotice>(), 2);
    ASSERT_EQ(listener1.Received<_Broker::StageContentsChanged>(), 1);
    ASSERT_EQ(listener1.Received<_Broker::ObjectsChanged>(), 1);
    ASSERT_EQ(listener1.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::LayerMutingChanged>(), 0);

    ASSERT_EQ(listener2.Received<_USD::StageNotice>(), 2);
    ASSERT_EQ(listener2.Received<_USD::StageContentsChanged>(), 1);
    ASSERT_EQ(listener2.Received<_USD::ObjectsChanged>(), 1);
    ASSERT_EQ(listener2.Received<_USD::StageEditTargetChanged>(), 0);
    ASSERT_EQ(listener2.Received<_USD::LayerMutingChanged>(), 0);
}

TEST(AddPrims, NoticesComparison)
{
    auto stage = PXR_NS::UsdStage::CreateInMemory();
    auto broker = PXR_NS::NoticeBroker::Create(stage);

    namespace _Broker = PXR_NS::UsdBrokerNotice;
    using _USD = PXR_NS::UsdNotice;

    // Subscribe to notices emitted by the broker.
    ::Test::Listener<
        _Broker::StageNotice,
        _Broker::StageContentsChanged,
        _Broker::ObjectsChanged,
        _Broker::StageEditTargetChanged,
        _Broker::LayerMutingChanged
    > listener1(stage);

    // Subscribe to Usd Notices emitted by the stage.
    ::Test::Listener<
        _USD::StageNotice,
        _USD::StageContentsChanged,
        _USD::ObjectsChanged,
        _USD::StageEditTargetChanged,
        _USD::LayerMutingChanged
    > listener2(stage);

    // Edit the stage...    
    stage->DefinePrim(PXR_NS::SdfPath {"/Foo"});
    stage->DefinePrim(PXR_NS::SdfPath {"/Bar"});
    stage->DefinePrim(PXR_NS::SdfPath {"/Baz"});

    // Ensure that similar notices are received via the stage and the broker. 
    ASSERT_EQ(listener1.Received<_Broker::StageNotice>(), 6);
    ASSERT_EQ(listener1.Received<_Broker::StageContentsChanged>(), 3);
    ASSERT_EQ(listener1.Received<_Broker::ObjectsChanged>(), 3);
    ASSERT_EQ(listener1.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::LayerMutingChanged>(), 0);

    ASSERT_EQ(listener2.Received<_USD::StageNotice>(), 6);
    ASSERT_EQ(listener2.Received<_USD::StageContentsChanged>(), 3);
    ASSERT_EQ(listener2.Received<_USD::ObjectsChanged>(), 3);
    ASSERT_EQ(listener2.Received<_USD::StageEditTargetChanged>(), 0);
    ASSERT_EQ(listener2.Received<_USD::LayerMutingChanged>(), 0);
}

TEST(MuteLayer, NoticesComparison)
{
    auto stage = ::Test::CreateStageWithLayers();
    auto broker = PXR_NS::NoticeBroker::Create(stage);

    namespace _Broker = PXR_NS::UsdBrokerNotice;
    using _USD = PXR_NS::UsdNotice;

    // Subscribe to notices emitted by the broker.
    ::Test::Listener<
        _Broker::StageNotice,
        _Broker::StageContentsChanged,
        _Broker::ObjectsChanged,
        _Broker::StageEditTargetChanged,
        _Broker::LayerMutingChanged
    > listener1(stage);

    // Subscribe to Usd Notices emitted by the stage.
    ::Test::Listener<
        _USD::StageNotice,
        _USD::StageContentsChanged,
        _USD::ObjectsChanged,
        _USD::StageEditTargetChanged,
        _USD::LayerMutingChanged
    > listener2(stage);

    // Edit the stage...    
    auto layers = stage->GetRootLayer()->GetSubLayerPaths();
    stage->MuteLayer(layers[0]);

    // Ensure that similar notices are received via the stage and the broker. 
    ASSERT_EQ(listener1.Received<_Broker::StageNotice>(), 3);
    ASSERT_EQ(listener1.Received<_Broker::StageContentsChanged>(), 1);
    ASSERT_EQ(listener1.Received<_Broker::ObjectsChanged>(), 1);
    ASSERT_EQ(listener1.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::LayerMutingChanged>(), 1);

    ASSERT_EQ(listener2.Received<_USD::StageNotice>(), 3);
    ASSERT_EQ(listener2.Received<_USD::StageContentsChanged>(), 1);
    ASSERT_EQ(listener2.Received<_USD::ObjectsChanged>(), 1);
    ASSERT_EQ(listener2.Received<_USD::StageEditTargetChanged>(), 0);
    ASSERT_EQ(listener2.Received<_USD::LayerMutingChanged>(), 1);
}

TEST(MuteLayers, NoticesComparison)
{
    auto stage = ::Test::CreateStageWithLayers();
    auto broker = PXR_NS::NoticeBroker::Create(stage);

    namespace _Broker = PXR_NS::UsdBrokerNotice;
    using _USD = PXR_NS::UsdNotice;

    // Subscribe to notices emitted by the broker.
    ::Test::Listener<
        _Broker::StageNotice,
        _Broker::StageContentsChanged,
        _Broker::ObjectsChanged,
        _Broker::StageEditTargetChanged,
        _Broker::LayerMutingChanged
    > listener1(stage);

    // Subscribe to Usd Notices emitted by the stage.
    ::Test::Listener<
        _USD::StageNotice,
        _USD::StageContentsChanged,
        _USD::ObjectsChanged,
        _USD::StageEditTargetChanged,
        _USD::LayerMutingChanged
    > listener2(stage);

    // Edit the stage...    
    auto layers = stage->GetRootLayer()->GetSubLayerPaths();

    // Keep ref pointer to the layer we try to mute and unmute to
    // prevent it for being destroyed when it is muted.
    auto _layer = PXR_NS::SdfLayer::FindOrOpen(layers[1]);

    stage->MuteLayer(layers[0]);
    stage->MuteLayer(layers[1]);
    stage->UnmuteLayer(layers[1]);
    stage->MuteAndUnmuteLayers(
        std::vector<std::string>{layers[2], layers[1]}, 
        std::vector<std::string>{});

    // Ensure that similar notices are received via the stage and the broker. 
    ASSERT_EQ(listener1.Received<_Broker::StageNotice>(), 12);
    ASSERT_EQ(listener1.Received<_Broker::StageContentsChanged>(), 4);
    ASSERT_EQ(listener1.Received<_Broker::ObjectsChanged>(), 4);
    ASSERT_EQ(listener1.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::LayerMutingChanged>(), 4);

    ASSERT_EQ(listener2.Received<_USD::StageNotice>(), 12);
    ASSERT_EQ(listener2.Received<_USD::StageContentsChanged>(), 4);
    ASSERT_EQ(listener2.Received<_USD::ObjectsChanged>(), 4);
    ASSERT_EQ(listener2.Received<_USD::StageEditTargetChanged>(), 0);
    ASSERT_EQ(listener2.Received<_USD::LayerMutingChanged>(), 4);
}

TEST(ChangeEditTarget, NoticesComparison)
{
    auto stage = ::Test::CreateStageWithLayers();
    auto broker = PXR_NS::NoticeBroker::Create(stage);

    namespace _Broker = PXR_NS::UsdBrokerNotice;
    using _USD = PXR_NS::UsdNotice;

    // Subscribe to notices emitted by the broker.
    ::Test::Listener<
        _Broker::StageNotice,
        _Broker::StageContentsChanged,
        _Broker::ObjectsChanged,
        _Broker::StageEditTargetChanged,
        _Broker::LayerMutingChanged
    > listener1(stage);

    // Subscribe to Usd Notices emitted by the stage.
    ::Test::Listener<
        _USD::StageNotice,
        _USD::StageContentsChanged,
        _USD::ObjectsChanged,
        _USD::StageEditTargetChanged,
        _USD::LayerMutingChanged
    > listener2(stage);

    // Edit the stage...
    auto layers = stage->GetRootLayer()->GetSubLayerPaths();
    auto layer1 = PXR_NS::SdfLayer::Find(layers[0]);
    auto layer2 = PXR_NS::SdfLayer::Find(layers[1]);

    stage->SetEditTarget(PXR_NS::UsdEditTarget(layer1));
    stage->SetEditTarget(PXR_NS::UsdEditTarget(layer2));

    // Ensure that similar notices are received via the stage and the broker. 
    ASSERT_EQ(listener1.Received<_Broker::StageNotice>(), 2);
    ASSERT_EQ(listener1.Received<_Broker::StageContentsChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::ObjectsChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::StageEditTargetChanged>(), 2);
    ASSERT_EQ(listener1.Received<_Broker::LayerMutingChanged>(), 0);

    ASSERT_EQ(listener2.Received<_USD::StageNotice>(), 2);
    ASSERT_EQ(listener2.Received<_USD::StageContentsChanged>(), 0);
    ASSERT_EQ(listener2.Received<_USD::ObjectsChanged>(), 0);
    ASSERT_EQ(listener2.Received<_USD::StageEditTargetChanged>(), 2);
    ASSERT_EQ(listener2.Received<_USD::LayerMutingChanged>(), 0);
}

TEST(CustomNotices, Handling)
{
    auto stage = PXR_NS::UsdStage::CreateInMemory();
    auto broker = PXR_NS::NoticeBroker::Create(stage);

    namespace _Broker = PXR_NS::UsdBrokerNotice;

    // Subscribe to notices emitted by the broker.
    ::Test::Listener<
        ::Test::MergeableNotice,
        ::Test::UnMergeableNotice
    > listener(stage);

    broker->Send<::Test::MergeableNotice>(::Test::DataMap({{"Foo", "Test1"}}));
    broker->Send<::Test::MergeableNotice>(::Test::DataMap({{"Foo", "Test2"}}));
    broker->Send<::Test::MergeableNotice>(::Test::DataMap({{"Bar", "Test3"}}));

    broker->Send<::Test::UnMergeableNotice>();
    broker->Send<::Test::UnMergeableNotice>();
    broker->Send<::Test::UnMergeableNotice>();

    ASSERT_EQ(listener.Received<::Test::MergeableNotice>(), 3);
    ASSERT_EQ(listener.Received<::Test::UnMergeableNotice>(), 3);
}

int main(int argc, char** argv)
{
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: test/python/test_broker.py
# :coding: utf-8

from pxr import Usd, Sdf, Tf
from usd_notice_broker import NoticeBroker, BrokerNotice

import pytest


def test_create():
    """Create a broker from stage."""
    stage = Usd.Stage.CreateInMemory()
    broker = NoticeBroker.Create(stage)
    assert broker.GetStage() == stage
    assert broker.IsInTransaction() is False

def test_create_twice():
    """Create two brokers from stage."""
    stage = Usd.Stage.CreateInMemory()
    broker1 = NoticeBroker.Create(stage)
    broker2 = NoticeBroker.Create(stage)
    assert broker1 == broker2

@pytest.mark.parametrize("notice_type, excepted", [
    ("StageNotice", 2),
    ("StageContentsChanged", 1),
    ("ObjectsChanged", 1),
    ("StageEditTargetChanged", 0),
    ("LayerMutingChanged", 0),
], ids=[
    "StageNotice",
    "StageContentsChanged",
    "ObjectsChanged",
    "StageEditTargetChanged",
    "LayerMutingChanged",
])
def test_add_prim(notice_type, excepted):
    """Add one prim to the stage and listen to notices.
    """
    stage = Usd.Stage.CreateInMemory()
    NoticeBroker.Create(stage)

    # Listen to broker notice.
    received_broker = []
    key1 = Tf.Notice.Register(
        getattr(BrokerNotice, notice_type),
        lambda n, _: received_broker.append(n), stage)

    # Listen to corresponding USD notice.
    received_usd = []
    key2 = Tf.Notice.Register(
        getattr(Usd.Notice, notice_type),
        lambda n, _: received_usd.append(n), stage)

    # Edit the stage...
    stage.DefinePrim("/Foo")

    # Ensure that we received the same number of notices.
    assert len(received_broker) == excepted
    assert len(received_usd) == excepted

@pytest.mark.parametrize("notice_type, excepted", [
    ("StageNotice", 6),
    ("StageContentsChanged", 3),
    ("ObjectsChanged", 3),
    ("StageEditTargetChanged", 0),
    ("LayerMutingChanged", 0),
], ids=[
    "StageNotice",
    "StageContentsChanged",
    "ObjectsChanged",
    "StageEditTargetChanged",
    "LayerMutingChanged",
])
def test_add_prims(notice_type, excepted):
    """Add several prims to the stage and listen to notices.
    """
    stage = Usd.Stage.CreateInMemory()
    NoticeBroker.Create(stage)

    # Listen to broker notice.
    received_broker = []
    key1 = Tf.Notice.Register(
        getattr(BrokerNotice, notice_type),
        lambda n, _: received_broker.append(n), stage)

    # Listen to corresponding USD notice.
    received_usd = []
    key2 = Tf.Notice.Register(
        getattr(Usd.Notice, notice_type),
        lambda n, _: received_usd.append(n), stage)

    # Edit the stage...
    stage.DefinePrim("/Foo")
    stage.DefinePrim("/Bar")
    stage.DefinePrim("/Baz")

    # Ensure that we received the same number of notices.
    assert len(received_broker) == excepted
    assert len(received_usd) == excepted

@pytest.mark.parametrize("notice_type, excepted", [
    ("StageNotice", 3),
    ("StageContentsChanged", 1),
    ("ObjectsChanged", 1),
    ("StageEditTargetChanged", 0),
    ("LayerMutingChanged", 1),
], ids=[
    "StageNotice",
    "StageContentsChanged",
    "ObjectsChanged",
    "StageEditTargetChanged",
    "LayerMutingChanged",
])
def test_mute_layer(notice_type, excepted, stage_with_layers):
    """Mute one layer and listen to notices.
    """
    stage = stage_with_layers
    NoticeBroker.Create(stage)

    # Listen to broker notice.
    received_broker = []
    key1 = Tf.Notice.Register(
        getattr(BrokerNotice, notice_type),
        lambda n, _: received_broker.append(n), stage)

    # Listen to corresponding USD notice.
    received_usd = []
    key2 = Tf.Notice.Register(
        getattr(Usd.Notice, notice_type),
        lambda n, _: received_usd.append(n), stage)

    # Edit the stage...
    layers = stage.GetRootLayer().subLayerPaths
    stage.MuteLayer(layers[0])

    # Ensure that we received the same number of notices.
    assert len(received_broker) == excepted
    assert len(received_usd) == excepted

@pytest.mark.parametrize("notice_type, excepted", [
    ("StageNotice", 12),
    ("StageContentsChanged", 4),
    ("ObjectsChanged", 4),
    ("StageEditTargetChanged", 0),
    ("LayerMutingChanged", 4),
], ids=[
    "StageNotice",
    "StageContentsChanged",
    "ObjectsChanged",
    "StageEditTargetChanged",
    "LayerMutingChanged",
])
def test_mute_layers(notice_type, excepted, stage_with_layers):
    """Mute several layers and listen to notices.
    """
    stage = stage_with_layers
    NoticeBroker.Create(stage)

    # Listen to broker notice.
    received_broker = []
    key1 = Tf.Notice.Register(
        getattr(BrokerNotice, notice_type),
        lambda n, _: received_broker.append(n), stage)

    # Listen to corresponding USD notice.
    received_usd = []
    key2 = Tf.Notice.Register(
        getattr(Usd.Notice, notice_type),
        lambda n, _: received_usd.append(n), stage)

    # Edit the stage...
    layers = stage.GetRootLayer().subLayerPaths

    # Keep ref pointer to the layer we try to mute and unmute to
    # prevent it for being destroyed when it is muted.
    _layer = Sdf.Layer.FindOrOpen(layers[1])

    stage.MuteLayer(layers[0])
    stage.MuteLayer(layers[1])
    stage.UnmuteLayer(layers[1])
    stage.MuteAndUnmuteLayers([layers[2], layers[1]], [])

    # Ensure that we received the same number of notices.
    assert len(received_broker) == excepted
    assert len(received_usd) == excepted

@pytest.mark.parametrize("notice_type, excepted", [
    ("StageNotice", 2),
    ("StageContentsChanged", 0),
    ("ObjectsChanged", 0),
    ("StageEditTargetChanged", 2),
    ("LayerMutingChanged", 0),
], ids=[
    "StageNotice",
    "StageContentsChanged",
    "ObjectsChanged",
    "StageEditTargetChanged",
    "LayerMutingChanged",
])
def test_change_edit_target(notice_type, excepted, stage_with_layers):
    """Change edit target and listen to notices.
    """
    stage = stage_with_layers
    NoticeBroker.Create(stage)

    # Listen to broker notice.
    received_broker = []
    key1 = Tf.Notice.Register(
        getattr(BrokerNotice, notice_type),
        lambda n, _: received_broker.append(n), stage)

    # Listen to corresponding USD notice.
    received_usd = []
    key2 = Tf.Notice.Register(
        getattr(Usd.Notice, notice_type),
        lambda n, _: received_usd.append(n), stage)

    # Edit the stage...
    layers = stage.GetRootLayer().subLayerPaths
    layer1 = Sdf.Layer.Find(layers[0])
    layer2 = Sdf.Layer.Find(layers[1])

    stage.SetEditTarget(Usd.EditTarget(layer1))
    stage.SetEditTarget(Usd.EditTarget(layer2))

    # Ensure that we received the same number of notices.
    assert len(received_broker) == excepted
    assert len(received_usd) == excepted
